### ISSUE TITLE
Official Docker Image

### ISSUE THREAD

**Issue by Gabology** [Caiyeon/goldfish#190]
Currently there is a `Dockerfile` in the repository but there is no official distribution of this image on Docker Hub.

We are also maintaining one under `themobilelife/goldfish` but I think it would be better if there was an official one maintained on Docker Hub.

**Comment by Caiyeon** [comment]
For the record, I agree. I just haven't found the time to test this end to end, and I don't think I should label a dockerfile on the hub "official" if I don't test it.

**Comment by ecliptik** [comment]
Should note that after troubleshooting an issue in #217, running Goldfish in a container may prove a challenge since if it's running in orchestration (k8s, Rancher, etc) and the container is rescheduled (host goes away, resources are re-allocated) it would be restarted with an expired wrapping token, and would lead to the service not coming up properly.

Some options are,

- set a longer ttl on the wrapping token (not a good idea)
- use an `entrypoint.sh` and have it include another app token to get a wrapping token on container restart (hacky)
- integrate with a secret bridge or something else with in the orchestration framework (probably the best idea)

**Comment by Caiyeon** [comment]
There are 2 other ways to do this.

1. Upon starting goldfish without a `token` cmd arg, there is an API open to receiving a wrapped token. So if there is a lifecycle hook (e.g. on k8s), a POST request with a newly generated wrapping token could be sent to goldfish to complete the process upon rescheduling. However, goldfish will lose all memory of what it previously received from requests and such.

2. There is a way to provide a raw token to goldfish (not wrapped, not approle, just regular token) via filename. This was implemented because nomad only offers a token file for vault integration. This is an option for persisting through restarts, but only if the file itself can be guaranteed to not be compromised. That is, anyone that can see the token will be able to theoretically access unseal keys given to goldfish.

I'm not a fan of the second method, but I implemented some time ago because it's the only thing Nomad supports. If I can come up with a big enough warning flag, the cmd arg may be renamed in future versions so that it can be used more widely.

Deployment mechanisms like terraform should make this easy to handle, by integrating with vault and injecting new wrapping tokens into template files. However, terraform would not solve the AWS spot instance problem.

Devops is hard.

**Comment by andrewsav-datacom** [comment]
Where is this API described? Cannot find anything on the Wiki.

Also: Vault allows you to pass configuration in json. Official vault docker image accepts the configuration in an environment variable, which entrypoint.sh writes to local.json for vault to use. Since hcl is inherently multi-line and goldfish does not seem to support json this makes this approach problematic.

If you want to make a truly reusable docker image, you should don't be baking your own docker.hcl in it, because I might not want yours, I might want my own, and having that without the need to build my own image is nice (less update issues),

Would you like me to open a separate issue to ask for json config support?

**Comment by andrewsav-datacom** [comment]
@ecliptik I understand the whole point of the wrapped token is that it is suppose to be provided manually, the same as vault unseal keys are meant to be provided manually for unseal. This, of course can be automated, but similar as hashicorp refuses to help with that, because in their view it defeats the point of the project similarly @Caiyeon might find it counter-productive to let you use unwrapped not time limited token. So if you are to implement your own solution, (that is without changes on Caiyeon's side ) your second bullet point is really the only option.

**Comment by Justin-DynamicD** [comment]
I'm on my phone ATM, but wanted to mention that it works perfectly with Nomad as the job will natively handle token renewal.  With new vault/K8s integration, there's likely a similar interaction.  Basically the scheduler handles token generation and shows it as a variable/file.

**Comment by Justin-DynamicD** [comment]
If also add that, at list in my view, the wrapped token holds the most value when used in custom schedulers, nomad itself it adds nothing as long as long as the native nomad ACL is in use (latest vault added a Nomad backed so you can request temporary management keys to submit jobs with which serves the exact same result)

**Comment by andrewsav-datacom** [comment]
Good point, however it would be nice if the cases that do not involve nomad or K8 could also be catered for. In this area there is a *lot* of variety of different setups and configs, and it pays to be careful about not blocking some useful scenarios.

**Comment by Justin-DynamicD** [comment]
I get what you're saying, I guess I'm concerned that it puts one down the path of putting the responsibility of accepting wrapped tokens and their renewal management afterwords as an app responsibility which would basically permanently limit adoption of vault.  While it's awesome goldfish is willing to go down this route, I think it makes sense to leave this portion _outside_ the responsibility of the app, and rather be handled by the orchestration/scheduling tools in use.  That just seems to be the more holistic solution to the problem.

Basically I'm not 100% sold that goldfish itself is the right place to be conquering this challenge, as you then still have to rinse repeat it for every container moving forward.  Again it's great it's there, but trying to bake wrapped secret handling into the container is going too far.

A generic goldfish container (which is basically the dockerfile described above) is all I use, I simply ship the hcl as an artifact in nomad and bang it works.  I would, however, gladly use an "official image" rather than my own custom one.

**Comment by aantono** [comment]
@Caiyeon Can you post a Nomad deployment example on the Wiki?

**Comment by Justin-DynamicD** [comment]
here's a generic example of what I use:

```
job "dev-goldfish" {
  region = "global"
  datacenters = ["dc1"]
  type = "service"
  update {
    stagger      = "30s"
    max_parallel = 1
  }
  group "goldfish" {
    count = 1
    restart {
      attempts = 10
      interval = "5m"
      delay    = "25s"
      mode     = "delay"
    }
    task "goldfish-container" {
      artifact {
        source      = "git::https://<internal github>/goldfish.git//goldfish"
        destination = "apps/goldfish"
      }
      driver = "docker"
      config {
        image = "elijahwright/goldfish"
        command = "/app/goldfish" 
        args = [
          "-config=/app/config/docker.hcl",
          "-nomad-token-file=${NOMAD_SECRETS_DIR}/vault_token",
        ]
        port_map {
          http = 8000
        }
        volumes = [
          "apps/goldfish:/app/config",
        ]
      }
      vault {
        policies = ["goldfish"]
      }
      service {
	      name = "goldfish"
        port = "http"
        tags = ["proxy"]
        check {
          type     = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }
      resources {
        cpu    = 100
        memory = 300 # in MB
        network {
          mbits = 100
          port "http" {}
        }
      }
    }
  }
}
```

Only thing "fancy" is the source artifact which is stating to copy the entire folder "goldfish" from the repository.  This contains the docker.hcl file which means the entire thing is stored in git.

**Comment by AndrewSav** [comment]
@Caiyeon ignore my question about json, I just found out that the hcl parser supports it natively, so this is a non-issue.

**Comment by Caiyeon** [comment]
The API is not documented. As much as I value documentation, time has been too tight in the past few months, and I would get no work done if I documented every HTTP API. 

The bootstrap API is just a post request to `<addr>/v1/bootstrap` with a body containing the key `wrapping_token`.

You can catch this request in the act by observing chrome network (debugging) tab while bootstrapping goldfish via the webpage. The request is sent upon completing this box: https://github.com/Caiyeon/goldfish/blob/master/screenshots/Bootstrap.png

This particular API I do believe should be documented, because the automation allows for a lot of machine-machine interactions. I'll add this to my to-dos.

**Comment by andrewsav-datacom** [comment]
Leaving aside the question if one should automatically unseal one's vault or if one should automatically bootstrap goldfish, I came up with a [script](https://github.com/andrewsav-datacom/vault-unseal) that does automatic unsealing bootstrapping.

With this view, if you want to automatically bootstrap goldfish, you can use something similar. But I would imagine in the majority of cases you *won't want* to bootstrap your goldfish the same as you won't want to automatically unseal your vault.

There are also use cases when you are running your vault in a container, and you already configured your goldfish for it, so when you start up a new goldfish container it should NOT do any kind of configuration.

With that in mind, I think that goldfish container should be minimal, e.g. something like this: https://github.com/andrewsav-datacom/docker-goldfish

The more minimalistic you go, the bigger number of useful scenarios you support. Just my two cents.

**Comment by adragoset** [comment]
Any chance we could get a version of the container that builds the source from a versioned branch. I'm trying to get goldfish to work with my vault behind client server ca's. I see the change to support vault ca's was made 4 days ago and is on master but the prebuilt binary was built 27 days ago.

**Comment by AndrewSav** [comment]
@adragoset in most cases that would mean that the changes are not ready to be published yet. When they are, a release is usually made.

**Comment by Caiyeon** [comment]
Binaries that are not built on specific versioned tags/commits are not recommended. While I can't guarantee bug-free releases, they are at least tested to a certain degree. Random commits have much less testing.

However, I will be adding building instructions in the wiki soon(TM) for those that wish to build their own binaries.

**Comment by Caiyeon** [comment]
Self-compiling instructions have been added to readme. The next upcoming release is ready and comes with slight changes to the deployment configuration file. 

Therefore, the docker file will have to wait until after this release. It's still coming, I'm just a bit overwhelmed with other work lately

**Comment by Caiyeon** [comment]
Experimental implementation: https://hub.docker.com/r/caiyeon/goldfish/

There may be bugs regarding ports and connectivity, but those will have to be fixed as they get reported. If anyone has improvements over how the docker implementation is done, feel free to open pull requests.

**Comment by AndrewSav** [comment]
Can you clarify, are you suggesting that everyone who needs different `config.hcl` than the one provided should rebuild the image themselves because they would not be able to use one from docker hub, because this config is baked in?

**Comment by Caiyeon** [comment]
The config is clearly not baked in. Please read the documentation.

**Comment by AndrewSav** [comment]
So the documentation clearly needs improvement. Because by just reading it it's not apparent how this is supposed to work. This is the request for explanation.

**Comment by AndrewSav** [comment]
@Caiyeon apologies, I misread it.

**Comment by Caiyeon** [comment]
This is step 2. I have no idea how I can possible make this any clearer than it is.

```
# 2. Create a config.hcl with your desired configuration (see wiki)
# set the file to an environment variable to be passed to docker later
export GOLDFISH_CONFIG=$(cat config.hcl)
```

**Comment by AndrewSav** [comment]
@Caiyeon I edited my comment above to clarify. 

Also I'm not completely sold on the EXPOSE part. since this something that comes from the config, it is not clear to me what benefit it has to put the port inside the image. After all when container is created by `docker run` command one can use the `--expose` flag to provide the port that matches to the one in their config if needed. But in my experience it is rarely needed at all.

**Comment by AndrewSav** [comment]
Just tested your image with my [sample installation](https://github.com/AndrewSav/test-vault) and it worked. Well done! ;)

Now I can change unofficial image to official one in that repo.
